### PR TITLE
feat(rust, python): Optional three-valued logic for any/all

### PR DESCRIPTION
**Pull request by magarick**
Resolves https://github.com/pola-rs/polars/issues/9472
Adds a drop_nulls parameter to allow null propagation in any and all.

### CODE REVIEW

**Comment by magarick** [comment]
The failing test looks unrelated. This should be ready to review.

**Comment by CameronBieganek** [comment]
I think that `any` and `all` should propagate `null` by default. I think by far the most reasonable definition for `any` is the following (using pseudocode for a functional language):

```
any([a, b, ..., z]) == ( a | b | ... | z )
```

By that definition, since `|` propagates nulls, `any` must also propagate nulls. Similar reasoning applies to `all`.

In fact, I don't think there should be a keyword argument at all. If one wants to exclude `null` before applying `any`, the idiom would be this:

```python
df.select(
    pl.col("x").drop_nulls().any()
)
```

...and similarly for `all`.

**Comment by magarick** [comment]
> I think that `any` and `all` should propagate `null` by default. I think by far the most reasonable definition for `any` is the following (using pseudocode for a functional language):
> 
> ```
> any([a, b, ..., z]) == ( a | b | ... | z )
> ```
> 
> By that definition, since `|` propagates nulls, `any` must also propagate nulls. Similar reasoning applies to `all`.

Then people will complain about changing default behavior. Introduce it first, give people time, then maybe change the default. It isn't at all obvious either that there's a universally "best" default for everyone and every use case.

> In fact, I don't think there should be a keyword argument at all. If one wants to exclude `null` before applying `any`, the idiom would be this:

Nah. Why have another function and why risk creating an intermediate result you don't need? The considerations are balancing speed, clarity, reasonable defaults, and reducing how much people have to type :-).

**Comment by CameronBieganek** [comment]
> Then people will complain about changing default behavior.

Polars is still in the 0.x series, so the API is expected to evolve.

> Nah. Why have another function

`drop_nulls()` already exists. Composability of `Expr` methods is better than using keyword arguments.

**Comment by magarick** [comment]
> > Then people will complain about changing default behavior.
> 
> Polars is still in the 0.x series, so the API is expected to evolve.

People will still have issues. This just happened with some changes I made to cut/qcut. Semver's not magic unfortunately.

> > Nah. Why have another function
> 
> `drop_nulls()` already exists. Composability of `Expr` methods is better than using keyword arguments.

I meant calling another function. The bigger issue would be guaranteeing it composed without creating another series along the way. Maybe it does, but I don't know enough about the internals yet.

**Comment by magarick** [comment]
@ritchie46 Is this OK? It doesn't change the default behavior and I think optionally propagating nulls is useful.

File: polars/polars-core/src/chunked_array/comparison/mod.rs [hunk 1]
```diff
@@ -1004,6 +1004,23 @@ impl BooleanChunked {
     pub fn any(&self) -> bool {
         self.downcast_iter().any(compute::boolean::any)
     }
+
+    // Three-valued versions which can return None
+    pub fn all_3val(&self, drop_nulls: bool) -> Option<bool> {
+        if drop_nulls || self.null_count() == 0 {
+            Some(self.downcast_iter().all(compute::boolean::all))
```
Comment (ritchie46) [inline polars/polars-core/src/chunked_array/comparison/mod.rs:1011]: Can simply call `self.all()` here

File: polars/polars-core/src/chunked_array/comparison/mod.rs [hunk 2]
```diff
@@ -1004,6 +1004,23 @@ impl BooleanChunked {
...
+        } else {
+            None
+        }
+    }
+    pub fn any_3val(&self, drop_nulls: bool) -> Option<bool> {
+        let res = self.downcast_iter().any(compute::boolean::any);
```
Comment (ritchie46) [inline polars/polars-core/src/chunked_array/comparison/mod.rs:1017]: we can directly call `self.any` here.

**Review by ritchie46 — COMMENTED** [review]
Looks good. One comment where we can reuse existing code.

File: polars/polars-core/src/chunked_array/comparison/mod.rs (hunk 1 above)
Comment (magarick) [reply polars/polars-core/src/chunked_array/comparison/mod.rs]: Yup. Done. Less code is better code.

**Comment by rob-p** [comment]
Hi all (cc @ritchie46) — I think this should be listed as a breaking change in the release notes for polars 0.32.0.  It adds a new required keyword argument to existing functions, and causes existing code using these functions to fail to compile.  It would be useful to list this as a breaking change, with a short description of what to pass to obtain the old (<= v0.31.1) behavior.

--Rob

**Pull request merged**